components/org.wso2.carbon.identity.query.saml/src/main/java/org/wso2/carbon/identity/query/saml/util/OpenSAML3Util.java

```
@@ -0,0 +1,316 @@
+        ((SAMLObjectContentReference) signature.getContentReferences().get(0)).setDigestAlgorithm(digestAlgorithm);
+
+        List<Signature> signatureList = new ArrayList<Signature>();
+        signatureList.add(signature);
+
+        MarshallerFactory marshallerFactory = XMLObjectProviderRegistrySupport.getMarshallerFactory();
+        Marshaller marshaller = marshallerFactory.getMarshaller(signableXMLObject);
+
+        try {
+            marshaller.marshall(signableXMLObject);
+        } catch (MarshallingException e) {
+            throw IdentityException.error("Unable to marshall the request", e);
+        }
+
+        Init.init();
+
+        try {
+            Signer.signObjects(signatureList);
+        } catch (SignatureException e) {
+            throw IdentityException.error("Error occurred while signing request", e);
+        }
+
+        return signableXMLObject;
+    }
+
+    /**
+     * Validate the signature of an assertion
+     *
+     * @param request    SAML Assertion, this could be either a SAML Request or a
+     *                   LogoutRequest
+     * @param alias      Certificate alias against which the signature is validated.
+     * @param domainName domain name of the subject
+     * @return true, if the signature is valid.
+     * @throws IdentityException If unable to load credential information
+     */
+    public static boolean validateXMLSignature(RequestAbstractType request, String alias,
+                                               String domainName) throws IdentityException {
+
+        boolean isSignatureValid = false;
+
+        if (request.getSignature() != null) {
+            try {
+                X509Credential cred = (X509Credential) OpenSAML3Util.getX509CredentialImplForTenant(domainName, alias);
+
+                SignatureValidator.validate(request.getSignature(), cred);
+                return true;
+            } catch (IdentitySAML2SSOException e) {
+                if (log.isDebugEnabled()) {
```

better to add a comment saying why this exception is ignored